### PR TITLE
add  partition optimizer status trainer

### PR DESCRIPTION
**Pull request by xinyual**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

**Comment by tgaddair** [comment]
This is awesome @xinyual! Can you also add a test?

**Comment by maxhgerlach** [comment]
This sounds intriguing! I guess a ZeRO-like optimizer would benefit from a ReduceScatter operation in Horovod (#1496)?

**Comment by leezu** [comment]
@maxhgerlach yes, though this PR only implements optimizer state sharding and not yet gradient sharding. Do you have any plans to revive #1496?

**Comment by maxhgerlach** [comment]
@leezu I was thinking about having a go at #1496 to get Reducescatter, but it would take a while. So if you need it earlier, please go ahead. :)

**Comment by tgaddair** [comment]
@xinyual, can you rebase your PR off master? The most recent changes should fix the failing buildkite and docs tests.

File: horovod/mxnet/partition_trainer/pos_trainer.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    def _allreduce_grads(self):
+        for i, param in enumerate(self._all_params):
+            if param.grad_req != 'null':
+                allreduce_(param.list_grad()[0], average=False,
+                           name=self._prefix + str(i), priority=-i,
+                           prescale_factor=1.0 / self._gradient_predivide_factor)
+
+    def step(self, batch_size, ignore_stale_grad=False):
+        """
+        inherit from trainer, only call boardcast to make sure all parameter are consistent
```
Comment (romerojosh) [inline horovod/mxnet/partition_trainer/pos_trainer.py:77]: Typo: boardcast -> broadcast

File: horovod/mxnet/partition_trainer/pos_trainer.py [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+
+import mxnet as mx
+from collections import OrderedDict, defaultdict
+import types
+import time
+import warnings
+from mxnet.gluon.parameter import Parameter
+
+
+class POS_Trainer(mx.gluon.Trainer):
```
Comment (romerojosh) [inline horovod/mxnet/partition_trainer/pos_trainer.py:14]: Can the name of this be more descriptive? Perhaps something like `PartitionedParameterTrainer` or similar?

Alternatively, do you think this parameter splitting code can be made a feature of the existing `hvd.DistributedTrainer` instead of creating a completely separate implementation?

File: horovod/mxnet/partition_trainer/pos_trainer.py [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+        self._partition_params, self._param2rank = self.partition_parameters(self._all_params)
+        self._own_part = self._partition_params[self._world_rank]
+        super(POS_Trainer, self).__init__(
+            self._own_part, optimizer, optimizer_params=optimizer_params, kvstore=None)
+        self._prefix = prefix if prefix else ""
+        self._scale = gradient_predivide_factor / size()
+        self._gradient_predivide_factor = gradient_predivide_factor
+
+
+    def partition_parameters(self, params):
```
Comment (romerojosh) [inline horovod/mxnet/partition_trainer/pos_trainer.py:51]: Add leading underscore to internal methods like this one. Also, add a brief comment describing what this does.

**Review by romerojosh — COMMENTED** [review]
Thanks for the PR. This looks very cool and would be a great addition to Horovod!

Left a few minor comments.

File: horovod/mxnet/partition_trainer/pos_trainer.py [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+            'is not supported. Try setting `update_on_kvstore` ' \
+            'to False when creating trainer.'
+
+        self._check_and_rescale_grad(self._scale / batch_size)
+        self._update(ignore_stale_grad)
+        self.broadcast_params()
+
+
+
+    def broadcast_params(self):
```
Comment (romerojosh) [inline horovod/mxnet/partition_trainer/pos_trainer.py:121]: This should also be renamed to something like `_broadcast_partitioned_params` to better differentiate it from the existing `_broadcast_parameters` method. Also, add a brief comment describing what it does.

**Comment by maxhgerlach** [comment]
@xinyual, @leezu: Are you still interested in partitioned, distributed optimizer states? I've just posted a revived PR #3299 that adds a ReduceScatter op, which might be helpful in this context.

**Comment by xinyual** [comment]
> @xinyual, @leezu: Are you still interested in partitioned, distributed optimizer states? I've just posted a revived PR #3299 that adds a ReduceScatter op, which might be helpful in this context.

I see that PR but I think it is not suitable. In the paper it mentions reduce-scatter but I think it is in model-level. For parameters, we may need to use reduce/allreduce. I open a new PR [#3309].(https://github.com/horovod/horovod/pull/3309) with testcase. Please see that.
